storage/cache/tile_set.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"errors"
+
+	"github.com/google/trillian/storage/storagepb"
+)
+
+// tileSet contains an append-only set of tree tiles.
+// TODO(pavelkalinnikov): Export it once it's ready for being used by storage.
+type tileSet struct {
+	tiles map[string]*storagepb.SubtreeProto
+}
+
+// newTileSet returns a new empty tileSet.
+func newTileSet() *tileSet {
+	return &tileSet{tiles: make(map[string]*storagepb.SubtreeProto)}
+}
+
+// Touch marks the passed-in tile ID as touched. Returns true iff it was not
+// touched before. Note that a tile can be touched but still not present, and
+// so the corresponding Get call can return nil.
+func (ts *tileSet) Touch(id subtreeID) bool {
```

Comment:
I like the other methods on this class, but I think the Touch method has some curious semantics that I think are brittle. Recommend moving the Touch method to something like a batchTileQueryBuilder (how Java does that sound?!), which would be something like:

batchTileQueryBuilder{
  getTile(subtreeID) { // add to set, no return type}
  fetchBatch(Transaction) tileset { ... }
}